packages/asset-swapper/src/utils/aggregate.ts

```diff
@@ -0,0 +1,800 @@
+        slippage,
+        isBuy,
+    );
+}
+
+function createBridgeOrder(
+    bridgeAddress: string,
+    makerToken: string,
+    takerToken: string,
+    makerAssetAmount: BigNumber,
+    takerAssetAmount: BigNumber,
+    slippage: number,
+    isBuy: boolean = false,
+): SignedOrderWithoutDomain {
+    return {
+        makerAddress: bridgeAddress,
+        makerAssetData: createBridgeAssetData(makerToken, bridgeAddress, createBridgeData(takerToken)),
+        takerAssetData: createERC20AssetData(takerToken),
+        ...createCommonOrderFields(makerAssetAmount, takerAssetAmount, slippage, isBuy),
+    };
+}
+
+/**
+ * Create ERC20Bridge asset data.
+ */
+export function createBridgeAssetData(tokenAddress: string, bridgeAddress: string, bridgeData: string): string {
+    const encoder = AbiEncoder.createMethod('ERC20Bridge', [
+        { name: 'tokenAddress', type: 'address' },
+        { name: 'bridgeAddress', type: 'address' },
+        { name: 'bridgeData', type: 'bytes' },
+    ]);
+    return encoder.encode({ tokenAddress, bridgeAddress, bridgeData });
+}
+
+/**
+ * Create ERC20Proxy asset data.
+ */
+export function createERC20AssetData(tokenAddress: string): string {
+    const encoder = AbiEncoder.createMethod('ERC20Token', [{ name: 'tokenAddress', type: 'address' }]);
+    return encoder.encode({ tokenAddress });
+}
+
+function createBridgeData(tokenAddress: string): string {
+    const encoder = AbiEncoder.create([{ name: 'tokenAddress', type: 'address' }]);
+    return encoder.encode({ tokenAddress });
+}
+
+type CommonOrderFields = Pick<
+    SignedOrderWithoutDomain,
+    Exclude<keyof SignedOrderWithoutDomain, 'makerAddress' | 'makerAssetData' | 'takerAssetData'>
+>;
+
+function createCommonOrderFields(
+    makerAssetAmount: BigNumber,
+    takerAssetAmount: BigNumber,
+    slippage: number,
+    isBuy: boolean = false,
+): CommonOrderFields {
+    return {
+        takerAddress: NULL_ADDRESS,
+        senderAddress: NULL_ADDRESS,
+        feeRecipientAddress: NULL_ADDRESS,
+        salt: generatePseudoRandomSalt(),
+        expirationTimeSeconds: INFINITE_TIMESTAMP_SEC,
+        makerFeeAssetData: NULL_BYTES,
+        takerFeeAssetData: NULL_BYTES,
+        makerFee: ZERO,
+        takerFee: ZERO,
+        makerAssetAmount: isBuy
+            ? makerAssetAmount
+            : makerAssetAmount.times(1 - slippage).integerValue(BigNumber.ROUND_UP),
+        takerAssetAmount: isBuy
+            ? takerAssetAmount.times(slippage + 1).integerValue(BigNumber.ROUND_UP)
+            : takerAssetAmount,
+        signature: '0x04',
+    };
+}
+
+// Get the partial output earned by a fill at input `partialInput`.
+function getPartialFillOutput(fill: Fill, partialInput: BigNumber): BigNumber {
+    return BigNumber.min(fill.output, fill.output.div(fill.input).times(partialInput)).integerValue(
+        BigNumber.ROUND_DOWN,
+    );
+}
+
+/**
+ * Class for finding optimized fill paths.
+ */
+export class FillsOptimizer {
```

Comment:
I know that AMM's like the one used by Uniswap will give you a worse price the more you fill. Is this the case for all of our sources?

Reply:
As a rule, yes. In practice, we often see quotes up to and beyond $10k, for high liquidity pairs, that are virtually linear (i.e., rates are essentially constant).

Reply:
Well, as a rule except for maybe Kyber, which is why we fall back to the flood fill.